File: web/src/lib/server/ssi.ts
const UA =
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0 Safari/537.36";

const CHART_URL = "https://iboard-api.ssi.com.vn/statistics/charts/history";

export type ChartData = {
  symbol: string;
  t: number[];
  c: number[];
};

export async function getChart(
  symbol: string,
  resolution = "5",
  lookbackSeconds = 259200
): Promise<ChartData> {
  const now = Math.floor(Date.now() / 1000);
  const from = now - lookbackSeconds;
  const url = `${CHART_URL}?resolution=${resolution}&symbol=${symbol.toUpperCase()}&from=${from}&to=${now}`;
  try {
    const r = await fetch(url, {
      headers: {
        "User-Agent": UA,
        "Accept": "application/json",
        "Origin": "https://iboard.ssi.com.vn",
        "Referer": "https://iboard.ssi.com.vn/",
      },
      next: { revalidate: 30 },
    });
    if (!r.ok) return { symbol: symbol.toUpperCase(), t: [], c: [] };
    const payload = await r.json();
    const data = payload?.data ?? {};
    return {
      symbol: symbol.toUpperCase(),
      t: data.t ?? [],
      c: data.c ?? [],
    };
  } catch {
    return { symbol: symbol.toUpperCase(), t: [], c: [] };
  }
}

File: api/sources/ssi.py
"""SSI iBoard chart history — intraday/daily OHLC for VN tickers + VNINDEX."""
from __future__ import annotations
import time
from .http import client

CHART_URL = "https://iboard-api.ssi.com.vn/statistics/charts/history"


async def get_chart(
    symbol: str, resolution: str = "5", lookback_seconds: int = 86_400
) -> dict:
    """Return {t,o,h,l,c,v} arrays for `symbol` over the lookback window.

    resolution: "1" | "5" | "15" | "30" | "60" | "1D"
    """
    now = int(time.time())
    frm = now - lookback_seconds
    r = await client().get(
        CHART_URL,
        params={"resolution": resolution, "symbol": symbol.upper(), "from": frm, "to": now},
    )
    r.raise_for_status()
    payload = r.json()
    data = payload.get("data") or {}
    return {
        "symbol": symbol.upper(),
        "resolution": resolution,
        "t": data.get("t") or [],
        "o": data.get("o") or [],
        "h": data.get("h") or [],
        "l": data.get("l") or [],
        "c": data.get("c") or [],
        "v": data.get("v") or [],
    }
